File: RESTFul/IMMN/Java/app1/src/main/java/com/att/api/webhooks/service/WebhooksService.java
/*
 * Copyright 2015 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.att.api.webhooks.service;

import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONException;

import com.att.api.oauth.OAuthToken;
import com.att.api.rest.APIResponse;
import com.att.api.rest.RESTClient;
import com.att.api.rest.RESTException;
import com.att.api.service.APIService;

import org.apache.commons.codec.EncoderException;
import org.apache.commons.codec.net.URLCodec;

/**
 * Used to interact with version 1 of the Webhooks API.
 *
 */
public class WebhooksService extends APIService {
    protected final URLCodec codec;

    /**
     * Creates a WebhooksService object.
     *
     * @param fqdn fully qualified domain name to use for sending requests
     * @param token OAuth token to use for authorization
     */
    public WebhooksService(String fqdn, OAuthToken token) {
        super(fqdn, token);
        codec = new URLCodec();
    }

    public CreateChannelResponse createNotificationChannel(
            final CreateChannelArgs args) throws RESTException {
        final String endpoint = getFQDN() + "/notification/v1/channels";

        final JSONObject jchannel = new JSONObject();
        jchannel.put("serviceName", args.getServiceName());
        jchannel.put("notificationContentType", args.getContentType());
        if (args.getVersion() != null)
            jchannel.put("notificationVersion", args.getVersion());
        final JSONObject jobj = new JSONObject();
        jobj.put("channel", jchannel);

        final APIResponse response = new RESTClient(endpoint)
                .addAuthorizationHeader(getToken())
                .setHeader("Accept", "application/json")
                .setHeader("Content-Type", "application/json")
                .setSuccessCode(201).httpPost(jobj.toString());

        final String location = response.getHeader("location");
        final String systemTransId = response
                .getHeader("x-systemTransactionId");

        try {
            JSONObject jobjresponse = new JSONObject(response.getResponseBody());
            JSONObject jchannelresponse = jobjresponse.getJSONObject("channel");
            final String channelId = jchannelresponse.getString("channelId");
            Integer maxEvts = null;
            if (jchannelresponse.has("maxEventsPerNotification")) {
                maxEvts = jchannelresponse
                        .getInt("maxEventsPerNotification");
            }

            final CreateChannelResponse.Channel channel = new CreateChannelResponse.Channel(
                    channelId, maxEvts);
            return new CreateChannelResponse(location, systemTransId, channel);
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public GetChannelResponse getNotificationChannelDetails(String channelId)
            throws RESTException {
        try {
            channelId = this.codec.encode(channelId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint
            = getFQDN() + "/notification/v1/channels/" + channelId;

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setHeader("Accept", "application/json")
            .setHeader("Content-Type", "application/json")
            .setSuccessCode(200)
            .httpGet();

        final String systemTransId = response.getHeader("x-systemTransactionId");
        try {
            JSONObject jobjresponse = new JSONObject(response.getResponseBody());
            JSONObject jchannelresponse = jobjresponse.getJSONObject("channel");
            final String channelIdResponse
                = jchannelresponse.getString("channelId");
            Integer maxEvts = null;
            if (jchannelresponse.has("maxEventsPerNotification")) {
                maxEvts = jchannelresponse.getInt("maxEventsPerNotification");
            }
            final String serviceName = jchannelresponse.getString("serviceName");
            String channelType = null;
            if (jchannelresponse.has("channelType")) {
                channelType = jchannelresponse.getString("channelType");
            }
            final String notificationContentType
                = jchannelresponse.getString("notificationContentType");
            final double notificationVersion
                = jchannelresponse.getDouble("notificationVersion");
            final GetChannelResponse.Channel channel
                = new GetChannelResponse.Channel(
                    channelIdResponse, maxEvts, serviceName, channelType,
                    notificationContentType, notificationVersion
                );
            return new GetChannelResponse(systemTransId, channel);
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public String deleteNotificationChannel(String channelId)
            throws RESTException {
        try {
            channelId = this.codec.encode(channelId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint
            = getFQDN() + "/notification/v1/channels/" + channelId;

        final APIResponse response = new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setSuccessCode(204)
            .httpDelete();

        return response.getHeader("x-systemTransactionId");
    }

    public CreateSubscriptionResponse createNotificationSubscription(
        final CreateSubscriptionArgs args
    ) throws RESTException {
        String channelId = args.getChannelId();
        try {
            channelId = this.codec.encode(channelId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint = getFQDN() + "/notification/v1/channels/"
            + channelId + "/subscriptions";

        final JSONObject jsubscription = new JSONObject()
            .put("events", args.getEvents());
        if (args.getCallbackData() != null) {
            jsubscription.put("callbackData", args.getCallbackData());
        }
        if (args.getExpiresIn() != null) {
            jsubscription.put("expiresIn", args.getExpiresIn());
        }

        final JSONObject jobj = new JSONObject()
            .put("subscription", jsubscription);

        final APIResponse response = new RESTClient(endpoint)
                .addAuthorizationHeader(getToken())
                .setHeader("Accept", "application/json")
                .setHeader("Content-Type", "application/json")
                .setSuccessCode(201)
                .httpPost(jobj.toString());

        return CreateSubscriptionResponse.valueOf(response);
    }

    public GetSubscriptionResponse getNotificationSubscriptionDetails(
        String channelId, String subscriptionId
    ) throws RESTException {
        try {
            channelId = this.codec.encode(channelId);
            subscriptionId = this.codec.encode(subscriptionId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint = getFQDN() + "/notification/v1/channels/"
            + channelId + "/subscriptions/" + subscriptionId;

        final APIResponse response = new RESTClient(endpoint)
                .addAuthorizationHeader(getToken())
                .setHeader("Accept", "application/json")
                .setSuccessCode(200)
                .httpGet();

        final String systemTransId = response
                .getHeader("x-systemTransactionId");
        final String contentType = response.getHeader("content-type");
    
        try {
            final JSONObject jobjresponse 
                = new JSONObject(response.getResponseBody());
            final JSONObject jsubscriptionresponse 
                = jobjresponse.getJSONObject("subscription");
            final String subscriptionIdResponse 
                = jsubscriptionresponse.getString("subscriptionId");
            final Integer expiresInResponse
                = jsubscriptionresponse.getInt("expiresIn");
            final JSONArray jevents =
                jsubscriptionresponse.getJSONArray("eventFilters");

            final String[] subscriptionEventsResponse 
                = new String[jevents.length()];
            for (int i = 0; i < jevents.length(); i++) {
                subscriptionEventsResponse[i] = jevents.getString(i);
            }
            final String callbackDataResponse
                = jsubscriptionresponse.getString("callbackData");

            final GetSubscriptionResponse.Subscription subscription
                = new GetSubscriptionResponse.Subscription(
                    subscriptionIdResponse, expiresInResponse,
                    subscriptionEventsResponse, callbackDataResponse
                );
                    
            return new GetSubscriptionResponse(
                contentType, systemTransId, subscription
            );
        } catch (JSONException pe) {
            throw new RESTException(pe);
        }
    }

    public UpdateSubscriptionResponse updateNotificationSubscription(
        final UpdateSubscriptionArgs args
    ) throws RESTException {
        String channelId = args.getChannelId();
        String subscriptionId = args.getSubscriptionId();
        try {
            channelId = this.codec.encode(channelId);
            subscriptionId = this.codec.encode(subscriptionId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint = getFQDN() + "/notification/v1/channels/"
            + channelId + "/subscriptions/" + subscriptionId;

        final JSONObject jsubscription = new JSONObject()
            .put("events", args.getEvents());
        if (args.getCallbackData() != null) {
            jsubscription.put("callbackData", args.getCallbackData());
        }
        if (args.getExpiresIn() != null) {
            jsubscription.put("expiresIn", args.getExpiresIn());
        }

        final JSONObject jobj = new JSONObject()
            .put("subscription", jsubscription);

        final APIResponse response = new RESTClient(endpoint)
                .addAuthorizationHeader(getToken())
                .setHeader("Accept", "application/json")
                .setHeader("Content-Type", "application/json")
                .setSuccessCode(200)
                .httpPut(jobj.toString());

        return UpdateSubscriptionResponse.valueOf(response);
    }

    /**
     * Returns x-systemTransactionId or null if successful
     *
     * @param channelId Channel id in use
     * @param subscriptionId Subscription id in use
     * @return String x-systemTransactionId or null
     *
     * @throws RESTException error happened making the request
     */
    public String deleteNotificationSubscription(
        String channelId, String subscriptionId
    ) throws RESTException {
        try {
            channelId = this.codec.encode(channelId);
            subscriptionId = this.codec.encode(subscriptionId);
        } catch (EncoderException e) {
            throw new RESTException(e);
        }
        final String endpoint = getFQDN() + "/notification/v1/channels/"
            + channelId + "/subscriptions/" + subscriptionId;
        return new RESTClient(endpoint)
            .addAuthorizationHeader(getToken())
            .setSuccessCode(204)
            .httpDelete()
            .getHeader("x-systemTransactionId");
    }
}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

File: RESTFul/IMMN/PHP/app1/lib/Webhooks/WebhooksService.php
<?php
namespace Att\Api\Webhooks;

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once __DIR__ . '../../Srvc/APIService.php';
require_once __DIR__ . '../../Srvc/ServiceException.php';
require_once __DIR__ . '/Channel.php';
require_once __DIR__ . '/CreateNotificationResponse.php';
require_once __DIR__ . '/CreateSubscriptionArgs.php';
require_once __DIR__ . '/CreateSubscriptionResponse.php';
require_once __DIR__ . '/GetNotificationResponse.php';
require_once __DIR__ . '/GetSubscriptionResponse.php';
require_once __DIR__ . '/UpdateSubscriptionArgs.php';
require_once __DIR__ . '/UpdateSubscriptionResponse.php';

use Att\Api\OAuth\OAuthToken;
use Att\Api\Restful\HttpGet;
use Att\Api\Restful\HttpPost;
use Att\Api\Restful\HttpPut;
use Att\Api\Restful\RestfulRequest;
use Att\Api\Srvc\APIService;
use Att\Api\Srvc\Service;
use Att\Api\Srvc\ServiceException;
use Att\Api\Webhooks\CreateSubscriptionArgs;
use Att\Api\Webhooks\CreateSubscriptionResponse;
use Att\Api\Webhooks\GetChannelResponse;
use Att\Api\Webhooks\GetNotificationResponse;
use Att\Api\Webhooks\GetSubscriptionResponse;
use Att\Api\Webhooks\SubscriptionResponse;
use Att\Api\Webhooks\UpdateSubscriptionArgs;
use Att\Api\Webhooks\UpdateSubscriptionResponse;

/**
 * Used to interact with version 1 of the Webhooks API.
 *
 * @category API
 * @package  Webhooks
 * @author   pk9069
 * @license  http://www.apache.org/licenses/LICENSE-2.0
 * @version  Release: @package_version@
 */
class WebhooksService extends APIService
{

    /**
     * Creates a WebhooksService object that can be used to interact with
     * the Webhooks API.
     *
     * @param string     $FQDN  fully qualified domain name to which requests
     *                          will be sent
     * @param OAuthToken $token OAuth token used for authorization
     */
    public function __construct($FQDN, OAuthToken $token)
    {
        parent::__construct($FQDN, $token);
    }

    public function createNotificationChannel(Channel $channel)
    {
        $endpoint = $this->getFqdn() . '/notification/v1/channels';

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json')
            ->setHeader('Content-Type', 'application/json');
        // PHP strips out .0 from 1.0 during json_encode; therefore, the string
        // has to be manually constructed.
        // Issue has been fixed in future PHP versions by specifying the
        // JSON_PRESERVE_ZERO_FRACTION flag.
        // See: https://bugs.php.net/bug.php?id=50224
        $bodyString = '{"channel":{"serviceName":';
        $bodyString .= ('"'. $channel->getServiceName() . '"');
        $bodyString .= ',"notificationContentType":';
        $bodyString .= json_encode($channel->getNotificationContentType());
        $bodyString .= ',"notificationVersion":1.0}}';
        $httpPost = new HttpPost();
        $httpPost->setBody($bodyString);
        $result = $req->sendHttpPost($httpPost);
        $location = $result->getHeader('location');
        $systemTransId = $result->getHeader('x-systemTransactionId');

        $successCodes = array(201);
        $arr = Service::parseJson($result, $successCodes);
        $arrChannel = $arr['channel'];
        $channelResponseId = $arrChannel['channelId'];
        $channelResponseMaxEvts = null;
        if (isset($arrChannel['maxEventsPerNotification'])) {
            $channelResponseMaxEvts = $arrChannel['maxEventsPerNotification'];
        }

        $channelResponse = new ChannelResponse(
            $channelResponseId, $channelResponseMaxEvts
        );

        return new CreateNotificationResponse(
            $location, $systemTransId, $channelResponse
        );
    }

    public function getNotificationChannel($channelId)
    {
        $channelId = urlencode($channelId);
        $suburl = '/notification/v1/channels/' . $channelId;
        $endpoint = $this->getFqdn() . $suburl;

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json');
        $result = $req->sendHttpGet();
        $systemTransId = $result->getHeader('x-systemTransactionId');

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);
        $arrChannel = $arr['channel'];
        $arrChannelId = $arrChannel['channelId'];
        $arrMaxEvts = null;
        if (isset($arrChannel['maxEventsPerNotification'])) {
            $arrMaxEvts = $arrChannel['maxEventsPerNotification'];
        }
        $arrServiceName = $arrChannel['serviceName'];
        $arrChannelType = null;
        if (isset($arrChannel['channelType'])) {
            $arrChannelType = $arrChannel['channelType'];
        }
        $arrNotificationContenType = $arrChannel['notificationContentType'];
        $arrNotificationVersion = $arrChannel['notificationVersion'];

        return new GetChannelResponse(
            $arrChannel, $arrMaxEvts, $arrServiceName, $arrChannelType,
            $arrNotificationContenType, $arrNotificationVersion
        );
    }

    public function deleteNotificationChannel($channelId)
    {
        $channelId = urlencode($channelId);
        $suburl = '/notification/v1/channels' . $channelId;
        $endpoint = $this->getFqdn() . $suburl;

        $req = new RestfulRequest($endpoint);
        $req->setAuthorizationHeader($this->getToken());
        $result = $req->sendHttpDelete();
        $code = $result->getResponseCode();
        $body = $result->getResponseBody();
        if ($code != 204) {
            throw new ServiceException($body, $code);
        }

        return $result->getHeader('x-systemTransactionId');
    }

    public function createNotificationSubscription(
        CreateSubscriptionArgs $args
    ) {
        $channelId = urlencode($args->getChannelId());
        $suburl = '/notification/v1/channels/' . $channelId . '/subscriptions';
        $endpoint = $this->getFqdn() . $suburl;

        $subscription = array("events" => $args->getEvents());
        if ($args->getCallbackData() != null) {
            $subscription['callbackData'] = $args->getCallbackData();
        }
        if ($args->getExpiresIn() != null) {
            $subscription['expiresIn'] = $args->getExpiresIn();
        }
        $jvals = json_encode(array("subscription" => $subscription));
        $httpPost = new HttpPost();
        $httpPost->setBody($jvals);

        $req = new RestfulRequest($endpoint);
        $result = $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Content-Type', 'application/json')
            ->setHeader('Accept', 'application/json')
            ->sendHttpPost($httpPost);

        $successCodes = array(201);
        $arr = Service::parseJson($result, $successCodes);
        $arrSubscription = $arr['subscription'];
        $arrSubscriptionId = $arrSubscription['subscriptionId'];
        $arrExpiresIn = null;
        if (isset($arrSubscription['expiresIn'])) {
            $arrExpiresIn = $arrSubscription['expiresIn'];
        }

        $subscriptionResponse = new SubscriptionResponse(
            $arrSubscriptionId, $arrExpiresIn
        );
        $contentType = $result->getHeader('content-type');
        $location = $result->getHeader('location');
        $systemTransId = $result->getHeader('x-systemTransactionId');
        $createSubscriptionResponse = new CreateSubscriptionResponse(
            $contentType, $location, $systemTransId, $subscriptionResponse
        );

        return $createSubscriptionResponse;
    }

    public function updateNotificationSubscription(
        UpdateSubscriptionArgs $args
    ) {
        $channelId = urlencode($args->getChannelId());
        $suburl = '/notification/v1/channels/' . $channelId . '/subscriptions/'
            . urlencode($args->getSubscriptionId());
        $endpoint = $this->getFqdn() . $suburl;

        $subscription = array("events" => $args->getEvents());
        if ($args->getCallbackData() != null) {
            $subscription['callbackData'] = $args->getCallbackData();
        }
        if ($args->getExpiresIn() != null) {
            $subscription['expiresIn'] = $args->getExpiresIn();
        }
        $jvals = json_encode(array("subscription" => $subscription));
        $httpPut = new HttpPut($jvals);

        $req = new RestfulRequest($endpoint);
        $result = $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Content-Type', 'application/json')
            ->setHeader('Accept', 'application/json')
            ->sendHttpPut($httpPut);

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);
        $arrSubscription = $arr['subscription'];
        $arrSubscriptionId = $arrSubscription['subscriptionId'];
        $arrExpiresIn = null;
        if (isset($arrSubscription['expiresIn'])) {
            $arrExpiresIn = $arrSubscription['expiresIn'];
        }

        $subscriptionResponse = new SubscriptionResponse(
            $arrSubscriptionId, $arrExpiresIn
        );
        $contentType = $result->getHeader('content-type');
        $systemTransId = $result->getHeader('x-systemTransactionId');
        $updateSubscriptionResponse = new UpdateSubscriptionResponse(
            $contentType, $systemTransId, $subscriptionResponse
        );

        return $updateSubscriptionResponse;
    }

    public function getNotificationSubscription($channelId, $subscriptionId)
    {
        $channelId = urlencode($channelId);
        $subscriptionId = urlencode($subscriptionId);
        $suburl = '/notification/v1/channels/' . $channelId . '/subscriptions/'
            . $subscriptionId;
        $endpoint = $this->getFqdn() . $suburl;

        $req = new RestfulRequest($endpoint);
        $req
            ->setAuthorizationHeader($this->getToken())
            ->setHeader('Accept', 'application/json');

        $result = $req->sendHttpGet();
        $contentType = $result->getHeader('content-type');
        $systemTransId = $result->getHeader('x-systemTransactionId');

        $successCodes = array(200);
        $arr = Service::parseJson($result, $successCodes);
        $arrSubscription = $arr['subscription'];
        $arrSubscriptionId = $arrSubscription['subscriptionId'];
        $arrExpiresIn = $arrSubscription['expiresIn'];
        /* TODO: remove work-around for events/eventFilters check */
        $arrEvents = null;
        if (isset($arrSubscription['events'])) {
            $arrEvents = $arrSubscription['events'];
        } else {
            $arrEvents = $arrSubscription['eventFilters'];
        }
        $arrCallbackData = $arrSubscription['callbackData'];

        return new GetSubscriptionResponse(
            $contentType, $arrSubscriptionId, $arrExpiresIn, $arrEvents,
            $arrCallbackData, $systemTransId
        );
    }

    public function deleteNotificationSubscription($channelId, $subscriptionId)
    {
        $channelId = urlencode($channelId);
        $subscriptionId = urlencode($subscriptionId);
        $suburl = '/notification/v1/channels/' . $channelId . '/subscriptions/';
        $suburl = $suburl . $subscriptionId;
        $endpoint = $this->getFqdn() . $suburl;

        $req = new RestfulRequest($endpoint);
        $req->setAuthorizationHeader($this->getToken());
        $result = $req->sendHttpDelete();
        $code = $result->getResponseCode();
        $body = $result->getResponseBody();
        if ($code != 204) {
            throw new ServiceException($body, $code);
        }

        return $result->getHeader('x-systemTransactionId');
    }
}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
?>
